File: python/src/pedro_agentware/llm/backend.py
"""LLM backend abstractions."""

from typing import TYPE_CHECKING, Protocol

from .request import Message

if TYPE_CHECKING:
    from .response import Response


class Backend(Protocol):
    """Protocol for LLM backends."""

    def complete(self, messages: list[Message]) -> "Response":
        """Send a conversation to the LLM and return its response."""
        ...

    def supports_native_tool_calling(self) -> bool:
        """Return True if backend supports native tool calling."""
        ...

    def model_name(self) -> str:
        """Return the model identifier."""
        ...

    def context_window_size(self) -> int:
        """Return the max tokens the model can process. 0 if unknown."""
        ...

File: go/llm/backend.go
package llm

import (
	"context"
)

// Backend is the abstraction over any OpenAI-compatible LLM API.
type Backend interface {
	// Complete sends a conversation to the LLM and returns its response.
	Complete(ctx context.Context, req *Request) (*Response, error)
	// SupportsNativeToolCalling returns true if the backend can handle
	// ToolDefinitions in the request and return structured ToolCalls.
	SupportsNativeToolCalling() bool
	// ModelName returns the model identifier being used.
	ModelName() string
	// ContextWindowSize returns the maximum number of tokens this backend
	// can process per request. Returns 0 if unknown.
	ContextWindowSize() int
}
